### PR TITLE
Add performance test for bulk subscribe over HTTP

### PR DESCRIPTION
**Pull request by shubham1172**
# Description

~**This PR should be merged only after** #5499 is merged and addressed in this PR.~ Completed.

This PR adds a k6 based test for comparing bulk subscribe HTTP API's performance with normal subscribe HTTP API's.

The test app is a simple go application that subscribes to routes using normal and bulk subscribe. The k6 test first populates the Kafka broker with a payload of 100 messages of size 100B each. The application listens for messages and pings back to the test app on getting 100 messages. This is done by multiple virtual actors.

Component used - Kafka.

Please see the comment for sample test results.

## Issue reference

Please reference the issue this PR will close: #5502 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by mcandeia** [comment]
[this](https://github.com/dapr/dapr/pull/5642) PR needs to be merged to fix the k6 operator version and then the perf tests will start to work again

**Comment by shubham1172** [comment]
> fyi: [this](https://github.com/dapr/dapr/pull/5642) PR needs to be merged to fix the k6 operator version and then the perf tests will start to work again

thanks @mcandeia, I have added the pinned version in this PR as well for now, I will merge master once the linked PR is merged.

File: tests/perf/pubsub_bulk_subscribe_http/test.js
Comment (mcandeia) [inline tests/perf/pubsub_bulk_subscribe_http/test.js:34]: this name will be used as the test identifier when exporting the metrics to somewhere,

I guess you can name it better, probably something like `bulkPSubscribe`

Comment (shubham1172) [reply tests/perf/pubsub_bulk_subscribe_http/test.js:34]: Thanks, updated!

File: tests/perf/pubsub_bulk_subscribe_http/test.js
Comment (mcandeia) [inline tests/perf/pubsub_bulk_subscribe_http/test.js:35]: fyi: https://k6.io/docs/using-k6/scenarios/executors/ you can try different types of executors

File: tests/perf/pubsub_bulk_subscribe_http/test.js
Comment (mcandeia) [inline tests/perf/pubsub_bulk_subscribe_http/test.js:95]: you can add a counter here and check that it should be zero (if not expected)

File: tests/perf/pubsub_bulk_subscribe_http/test.js
Comment (mcandeia) [inline tests/perf/pubsub_bulk_subscribe_http/test.js:98]: same here, use a custom counter and create a custom threshold

**Review by mcandeia — CHANGES_REQUESTED** [review]

File: tests/apps/perf/pubsub_subscribe_http/Dockerfile
Comment (shubham1172) [reply tests/apps/perf/pubsub_subscribe_http/Dockerfile:1]: The problem is that buildtools only builds `app.go` explicitly. My code is broken into multiple files, so it results in an error if I remove the Dockerfile. It becomes harder to read/maintain if I merge them into a single file. Build tools also had issues locating the default Dockerfile. I am fixing it in this PR. /cc @ItalyPaleAle to review.

File: tests/perf/pubsub_bulk_subscribe_http/test.js
Comment (shubham1172) [reply tests/perf/pubsub_bulk_subscribe_http/test.js:98]: Added!

File: tests/perf/pubsub_bulk_subscribe_http/test.js
Comment (shubham1172) [reply tests/perf/pubsub_bulk_subscribe_http/test.js:95]: The app closes the socket connection after writing the test response and triggers the `close` event. So it should not be zero. Also the k6 test continues to run until the app closes the socket (since we don't have an explicit `socket.Close()` or timeout on our end).

Comment (shubham1172) [reply tests/perf/pubsub_bulk_subscribe_http/test.js:95]: Added an explicit timeout and counter to check if we had to close manually.

File: tests/perf/pubsub_bulk_subscribe_http/pubsub_bulk_subscribe_http_test.go
Comment (shubham1172) [reply tests/perf/pubsub_bulk_subscribe_http/pubsub_bulk_subscribe_http_test.go]: It might be useful, I have added that option, thanks!

File: tests/perf/pubsub_bulk_subscribe_http/test.js
Comment (shubham1172) [reply tests/perf/pubsub_bulk_subscribe_http/test.js]: Thanks to this, TIL that Gorilla websocket does not close the connection gracefully by default. I had to update my app to send a special control message and close the websocket connection gracefully.

**Comment by shubham1172** [comment]
Adding this here for reference - 100 events of 100 bytes each with 500 virtual users (ramping-vus), published using bulk publish, given are various HTTP request duration in milliseconds - 

## Normal subscribe

```json
{
"type": "trend",
"contains": "time",
"values": {
"max": 1176.942004,
"p(90)": 986.711315,
"p(95)": 1066.0373890999997,
"avg": 502.74581646805916,
"min": 0.309318,
"med": 475.461188
}
```

## Bulk subscribe

```json
{
"type": "trend",
"contains": "time",
"values": {
"max": 470.397741,
"p(90)": 248.4924315,
"p(95)": 357.75256375000004,
"avg": 129.70715183285915,
"min": 0.556631,
"med": 105.968547
}
```

**Comment by shubham1172** [comment]
Adding this here for reference - 100 events of 100 bytes each with 500 virtual users (ramping-vus), published using bulk publish, given are various HTTP request duration in milliseconds - 

## Normal subscribe 

```json
{
"type": "trend",
"contains": "time",
"values": {
"max": 1176.942004,
"p(90)": 986.711315,
"p(95)": 1066.0373890999997,
"avg": 502.74581646805916,
"min": 0.309318,
"med": 475.461188
}
```

## Bulk subscribe

```json
{
"type": "trend",
"contains": "time",
"values": {
"max": 470.397741,
"p(90)": 248.4924315,
"p(95)": 357.75256375000004,
"avg": 129.70715183285915,
"min": 0.556631,
"med": 105.968547
}
```

File: tests/apps/perf/pubsub_subscribe_http/subscribe.go
Comment (mukundansundar) [inline tests/apps/perf/pubsub_subscribe_http/subscribe.go:69]: should message uniqueness be considered as part of perf? 
As in measuring all the events have been received (p99) vs there being some duplicate events?
probably publish a json event with a unique ID field and have a map/set of IDs seen?

Comment (shubham1172) [reply tests/apps/perf/pubsub_subscribe_http/subscribe.go:69]: > should message uniqueness be considered as part of perf? 
Isn't that a **correctness** test, which should be an e2e test scenario. WDYT? These checks will introduce further latency.

Comment (mukundansundar) [reply tests/apps/perf/pubsub_subscribe_http/subscribe.go:69]: But isn't perf supposed to measure complete response time only? And not partial which will happen on retries?

File: tests/perf/pubsub_bulk_subscribe_http/pubsub_bulk_subscribe_http_test.go
Comment (DeepanshuA) [inline tests/perf/pubsub_bulk_subscribe_http/pubsub_bulk_subscribe_http_test.go:35]: Is it used somewhere?

File: tests/apps/perf/pubsub_subscribe_http/app.go
Comment (DeepanshuA) [inline tests/apps/perf/pubsub_subscribe_http/app.go:34]: Though it seems fine to have a hard-coded number here, as we are running perf test with certain assumptions here. But still, do you want to make it configurable, in case we would like to run perf test in future with different benchmarks with different number of messages and sizes?

Comment (shubham1172) [reply tests/apps/perf/pubsub_subscribe_http/app.go:34]: Nice suggestion, made it configurable via environment variables.

File: tests/perf/pubsub_bulk_subscribe_http/pubsub_bulk_subscribe_http_test.go
Comment (shubham1172) [reply tests/perf/pubsub_bulk_subscribe_http/pubsub_bulk_subscribe_http_test.go:35]: Nope, just a leftover, removed, thanks!

**Review by DeepanshuA — DISMISSED** [review]

File: tests/apps/perf/pubsub_subscribe_http/subscribe.go
Comment (shubham1172) [reply tests/apps/perf/pubsub_subscribe_http/subscribe.go:69]: TODO: Only count unique messages in subscriber application.

**Review by artursouza — APPROVED** [review]

**Comment by artursouza** [comment]
Will merge after /ok-to-perf has concluded.

**Comment by shubham1172** [comment]
@artursouza FYI perf tests concluded.

**Review by DeepanshuA — APPROVED** [review]

**Pull request merged**
